### PR TITLE
[SPARK-40789][PYTHON][TESTS] Separate tests under pyspark.sql.tests

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to split the tests into the sub-packages:

**Before**

```
tests
├── __init__.py
├── test_arrow.py
├── test_arrow_map.py
├── test_catalog.py
├── test_column.py
├── test_conf.py
├── test_connect_basic.py
├── test_connect_column_expressions.py
├── test_connect_plan_only.py
├── test_connect_select_ops.py
├── test_context.py
├── test_dataframe.py
├── test_datasources.py
├── test_functions.py
├── test_group.py
├── test_pandas_cogrouped_map.py
├── test_pandas_grouped_map.py
├── test_pandas_grouped_map_with_state.py
├── test_pandas_map.py
├── test_pandas_udf.py
├── test_pandas_udf_grouped_agg.py
├── test_pandas_udf_scalar.py
├── test_pandas_udf_typehints.py
├── test_pandas_udf_typehints_with_future_annotations.py
├── test_pandas_udf_window.py
├── test_readwriter.py
├── test_serde.py
├── test_session.py
├── test_streaming.py
├── test_streaming_listener.py
├── test_types.py
├── test_udf.py
├── test_udf_profiler.py
├── test_utils.py
└── typing
    ├── ...
```

**After**

```
tests
├── __init__.py
├── connect
│   ├── __init__.py
│   ├── test_connect_basic.py
│   ├── test_connect_column_expressions.py
│   ├── test_connect_plan_only.py
│   └── test_connect_select_ops.py
├── pandas
│   ├── __init__.py
│   ├── test_pandas_cogrouped_map.py
│   ├── test_pandas_grouped_map.py
│   ├── test_pandas_grouped_map_with_state.py
│   ├── test_pandas_map.py
│   ├── test_pandas_udf.py
│   ├── test_pandas_udf_grouped_agg.py
│   ├── test_pandas_udf_scalar.py
│   ├── test_pandas_udf_typehints.py
│   ├── test_pandas_udf_typehints_with_future_annotations.py
│   └── test_pandas_udf_window.py
├── streaming
│   ├── __init__.py
│   ├── test_streaming.py
│   └── test_streaming_listener.py
├── test_arrow.py
├── test_arrow_map.py
├── test_catalog.py
├── test_column.py
├── test_conf.py
├── test_context.py
├── test_dataframe.py
├── test_datasources.py
├── test_functions.py
├── test_group.py
├── test_readwriter.py
├── test_serde.py
├── test_session.py
├── test_types.py
├── test_udf.py
├── test_udf_profiler.py
├── test_utils.py
└── typing
    ├── ...
```

This way is consistent with `pyspark.pandas.tests`.

### Why are the changes needed?

To make it easier to maintain, track and add the tests.

### Does this PR introduce _any_ user-facing change?

No, dev-only.

### How was this patch tested?

CI in this PR should test it out.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @zhengruifeng @ueshin @grundprinzip @xinrong-meng @itholic please take a look when you guys find some time to review.

**Review by grundprinzip — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

File: dev/sparktestsupport/modules.py [hunk 1]
```diff
@@ -461,22 +461,22 @@ def __hash__(self):
...
-        "pyspark.sql.tests.test_pandas_udf_grouped_agg",
-        "pyspark.sql.tests.test_pandas_udf_scalar",
-        "pyspark.sql.tests.test_pandas_udf_typehints",
-        "pyspark.sql.tests.test_pandas_udf_typehints_with_future_annotations",
-        "pyspark.sql.tests.test_pandas_udf_window",
+        "pyspark.sql.tests.pandas.test_pandas_cogrouped_map",
+        "pyspark.sql.tests.pandas.test_pandas_grouped_map",
+        "pyspark.sql.tests.pandas.test_pandas_grouped_map_with_state",
+        "pyspark.sql.tests.pandas.test_pandas_map",
+        "pyspark.sql.tests.pandas.test_arrow_map",
```
Comment (HyukjinKwon) [inline dev/sparktestsupport/modules.py:468]:
```suggestion
        "pyspark.sql.tests.test_arrow_map",
```

**Review by zhengruifeng — APPROVED** [review]
LGTM

**Review by itholic — APPROVED** [review]
+1. LGTM

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**